src/helper/rsa.rs

```diff
@@ -0,0 +1,205 @@
+
+use crate::{
+    bail,
+    crypto::{hash::HashAlgorithm, public_key::PublicKeyAlgorithm},
+    errors::Result,
+    packet::PublicKey,
+    types::{
+        EskType, Fingerprint, KeyId, KeyVersion, Mpi, PkeskBytes, PublicKeyTrait, PublicParams,
+        SecretKeyTrait, SignatureBytes, Version,
+    },
+};
+
+use super::{PgpHash, PgpPublicKey};
+
+impl PgpPublicKey for RsaPublicKey {
+    const PGP_ALGORITHM: PublicKeyAlgorithm = PublicKeyAlgorithm::RSA;
+
+    fn pgp_parameters(&self) -> PublicParams {
+        PublicParams::RSA {
+            n: self.n().into(),
+            e: self.e().into(),
+        }
+    }
+}
+
+/// [`signature::Signer`] backed signer for PGP.
+pub struct RsaSigner<T, D> {
+    inner: T,
+    public_key: PublicKey,
+    _digest: PhantomData<D>,
+}
+
+impl<D, T> fmt::Debug for RsaSigner<T, D> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RsaSigner").finish()
+    }
+}
+
+impl<T, D> RsaSigner<T, D>
+where
+    D: Digest,
+    T: Keypair<VerifyingKey = VerifyingKey<D>>,
+{
+    /// Create a new signer with a given public key
+    pub fn new(inner: T, created_at: DateTime<Utc>) -> Result<Self> {
+        let public_key = PublicKey::new(
+            Version::New,
+            KeyVersion::V4,
+            RsaPublicKey::PGP_ALGORITHM,
+            created_at,
+            None,
+            inner.verifying_key().as_ref().pgp_parameters(),
+        )?;
+
+        Ok(Self {
+            inner,
+            public_key,
+            _digest: PhantomData,
+        })
+    }
+}
+
+impl<D, T> RsaSigner<T, D>
+where
+    D: Digest + PgpHash,
+    T: PrehashSigner<Signature>,
+{
+    fn sign_prehash(&self, hash: HashAlgorithm, prehash: &[u8]) -> Result<Vec<Vec<u8>>> {
+        if D::HASH_ALGORITHM != hash {
+            bail!(
+                "Signer only support {expected:?}, found {found:?}",
+                expected = D::HASH_ALGORITHM,
+                found = hash
+            );
+        }
+        // TODO: check digest size
+        let sig = self.inner.sign_prehash(prehash)?;
+
+        Ok(vec![sig.to_vec()])
+    }
+}
+
+impl<D, T> Keypair for RsaSigner<T, D>
+where
+    T: Keypair,
+{
+    type VerifyingKey = T::VerifyingKey;
+
+    fn verifying_key(&self) -> Self::VerifyingKey {
+        self.inner.verifying_key()
+    }
+}
+
+impl<D, T> SecretKeyTrait for RsaSigner<T, D>
+where
+    T: PrehashSigner<Signature>,
+    D: Digest + PgpHash,
+{
+    type PublicKey = PublicKey;
+    type Unlocked = Self;
+
+    fn unlock<F, G, Tr>(&self, _pw: F, work: G) -> Result<Tr>
+    where
+        F: FnOnce() -> String,
+        G: FnOnce(&Self::Unlocked) -> Result<Tr>,
+    {
+        work(self)
+    }
+
+    fn create_signature<F>(
+        &self,
+        _key_pw: F,
+        hash: HashAlgorithm,
+        prehashed_data: &[u8],
+    ) -> Result<SignatureBytes>
+    where
+        F: FnOnce() -> String,
+    {
+        let sig = self.sign_prehash(hash, prehashed_data)?;
```

Comment:
There's quite a lot of repetition with this and the ECDSA case... I wonder if it's possible to somehow move it to a common function or something :thinking:

Reply:
I don't think I can do that, maybe a macro, but even then there are subtle differences and I worry about readability.